flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -374,13 +382,27 @@ private boolean triggerSnapshotIfNeeded(FlinkResourceContext<CR> ctx, SnapshotTy
         var resource = ctx.getResource();
         var conf = ctx.getObserveConfig();
 
-        Optional<SnapshotTriggerType> triggerOpt =
-                SnapshotUtils.shouldTriggerSnapshot(resource, conf, snapshotType);
+        var timestampStore =
+                snapshotType == SAVEPOINT
+                        ? savepointTriggerTimestamps
+                        : checkpointTriggerTimestamps;
+        Set<FlinkStateSnapshot> snapshots = Set.of();
+        if (FlinkStateSnapshotUtils.isSnapshotResourceEnabled(ctx.getOperatorConfig(), conf)) {
+            snapshots = ctx.getJosdkContext().getSecondaryResources(FlinkStateSnapshot.class);
+        }
+
+        var lastTrigger = timestampStore.getLastPeriodicTriggerInstant(resource, snapshots);
```

Comment:
feels like in almost all cases we don't actually need to get the secondary resource. As long as we have the periodic timestamp already available. Can we pass a supplier or have some other lazy logic?